### ISSUE TITLE
the trait bound `Self: std::marker::Sized` is not satisfied [E0277] false positive

### ISSUE THREAD

**Issue by lovasoa** [intellij-rust/intellij-rust#6991]
## Environment

* **IntelliJ Rust plugin version:** 0.3.143.3737-203
* **Rust toolchain version:** 1.50
* **IDE name and version:** IntelliJ

## Problem description

The following code raises an error when it shouldn't: 

<img width="772" alt="image" src="https://user-images.githubusercontent.com/552629/111756159-31fdf480-889a-11eb-8a42-59f41694307d.png">

The plugin ignores the where clause and thinks that `Self` is not sized

## Steps to reproduce

```rust
trait T {
    fn f(self) -> Self
    where
        Self: Sized,
    {
        self
    }
}
```

**Comment by qiujiangkun** [comment]
I have met the same problem recently.

**Comment by Seeker14491** [comment]
This looks like a duplicate of #2530.

**Comment by qiujiangkun** [comment]
It appears in the latest version. Maybe someone can do a bisection?

**Comment by afetisov** [comment]
Surprisingly, the bug doesn't happen if Self isn't the last generic parameter:

```rust
trait T {
    fn f(self) -> Foo<Self, u32> // no error
        where
            Self: Sized,
    {
        Foo(self, 0)
    }
}

struct Foo<A, B>(A, B);
```

```rust
trait T {
    fn f(self) -> Foo<u32, Self, u32> // no error
        where
            Self: Sized,
    {
        Foo(0, self, 0)
    }
}

struct Foo<A, B, C>(A, B, C);
```

But if Self is last, the error triggers:
```rust
trait T {
    fn f(self) -> Foo<u32, u32, Self> // error E0277
        where
            Self: Sized,
    {
        Foo(0, 0, self)
    }
}

struct Foo<A, B, C>(A, B, C);
```
